Prepare zip editor for apache commons-compress usage

mehmet-karaman opened the pull request:
- Prepare code for apache commons-compress usage.
-  update java version
- fix exception problem when zip file is not a zip file but the getNextEntry method was called.

mehmet-karaman committed:
Prepare code for apache commons-compress usage plain apache
commons-compress zip changes. update java version
fix exception problem when zip file is not a zip file but the
getNextEntry method was called.

uvoigt reviewed (changes requested):
Hi @mehmet-karaman 

For whatever reason, the size of an archive entry is no longer visible:

before:
<img width="380" height="201" alt="image" src="https://github.com/user-attachments/assets/9e53cd80-e581-4014-9a77-4b2b4a75a641" />

after:
<img width="462" height="295" alt="image" src="https://github.com/user-attachments/assets/233ed77c-5a9f-4ecd-8462-3c602334937b" />

This is because the Apache library is not able to evaluate the value of size while reading

**Note**: `ZipArchiveInputStream` may create entries that return `SIZE_UNKNOWN SIZE_UNKNOWN` as long as the entry hasn't been read completely.

What was the reason to replace the original Java implementation by the Apache implementation?

mehmet-karaman commented:
Hi, thanks I will check that.. If i recall correctly the apache commons compress was setting the sizes automatically (rewrites the header). Might be I missed something there..

mehmet-karaman commented:
Checked on my machine: 
<img width="660" height="450" alt="image" src="https://github.com/user-attachments/assets/0c3a867e-5dd4-4b98-8f9c-f37bfc69127e" />

Could it be that your archive is broken? Could you provide the archive, so i can check it in the hex editor.

uvoigt commented:
Please open a Jar-file written by a Jar task. You should see the behaviour.

mehmet-karaman commented:
Yes.. Jar files seem to have a problem.. I am going to check this.. Thanks..

mehmet-karaman committed:
Fix size problem of the entries with -1 size in file header and add file
problem

mehmet-karaman commented:
@uvoigt I've added a new test for jars (checking that the true size is written instead of -1) and fixed that bug, fixed also a bug in saving archives.. (the second PR contained already the fix, but I accidentally forgot it to pick the change into this PR) .. will rebase the second PR after this PR was merged.

iloveeclipse commented on ZipEditor/src/zipeditor/model/Node.java:
```diff
@@ -158,6 +160,13 @@ public void setTime(long time) {
+			try {
+				size = IOUtils.copyLarge(getContent(), NullOutputStream.INSTANCE);
+			} catch (IOException e) {
+				e.printStackTrace();
```
Proper logging?

mehmet-karaman replied:
:) realtime reactions.. Thanks.. I will fix this..

mehmet-karaman replied:
done.

uvoigt commented on ZipEditor/src/zipeditor/model/Node.java:
```diff
@@ -158,6 +162,13 @@ public void setTime(long time) {
 	}
 
 	public long getSize() {
```
We saw this occurring in Zip file (Jar files) only. `getSize` could be overridden in `ZipNode`.
For error logging  there is
`model.logError(e);`
e.g. within `getContent`

but I do not remember exactly why I applied this pattern. (I built an RCP app which I had never released to the public which may have profited from that feature.)

This is a comment only, no change request.

mehmet-karaman commented:
Thank you.. I am going to rebase the zstd support PR  to make it ready for review..

uvoigt commented:
@mehmet-karaman 
The commit 99b53e972e8da47db85ebf5ba957c298de08b046 clearly slows down the opening of a rather large zipfile. I was not exact enough when testing this PR.

Did I forget to ask: What was the particular reason to introduce Apache compress for zip compression in favour of the Java Zip implementation?

From a read performance point of view, commons-compress is not in a good position:
```
1.085.763 entries read in 0.096s using java.util.zip.ZipFile
1.085.763 entries read in 1.687s using java.util.zip.ZipInputStream
1.085.763 entries read in 12.919s using org.apache.commons.compress.archivers.zip.ZipArchiveInputStream
```

The fact that the entire file might be read only to display the size of an entry is even worse.

Within the ZipEditor, there are other mechanisms as listener notifications that make it run much slower.

Regards,
Uwe

mehmet-karaman commented:
@uvoigt It wasn't possible to override some functionalities in the zip implementation of java. Additionally, we would like to have Zstd support (the second PR is the zstd support impl) in the Zip Editor, which is only provided by apache commons-compress. 

What is the real problem there: To read the File size when size isn't set? Maybe we can improve it in that way? 

I would like to try to reproduce it on my own, can you tell me how to reproduce this?

uvoigt commented:
@mehmet-karaman 
I am okay with Zstd support and will approve the related PR.
But for Zip I probably prepare another branch that uses ZipFile at least if a file is available. Also, it changes notifying in the load process and applies other modifications which slow down the editor initialization.

The case which reproduces the behaviour is a large zip file with a million small entries.

It takes a few minutes but you can use the code to create one

```
		ZipOutputStream zip = new ZipOutputStream(new FileOutputStream("large.zip"));
		for (int i = 0; i < 1000000; i++) {
			ZipEntry entry = new ZipEntry(Math.random() * 1000000 + ".txt");
			zip.putNextEntry(entry);
			zip.write((Math.random() * 100000 + "").getBytes());
		}
		zip.close();

```

mehmet-karaman commented:
Thank you, I would like to investigate this. I will profile the process and see what can be done.

mehmet-karaman commented:
I've analyzed the performance degradation, and it appears to be irrelevant for my use case. We work with very large ZIP files that contain relatively few entries, so iterating over all entries does not introduce any noticeable overhead. However, I would still like to keep our branches aligned.

Do you have any suggestions on how we should proceed? Would you need support in implementing the Java ZIP integration and Apache Commons? The commons-compress ZIP implementation is more feature‑complete and handles many edge cases across ZIP files. Perhaps it could also serve as a fallback option?

uvoigt commented:
If there were something to detect if `java.util.zip.ZipFile` would have problems to open a file, I would suggest to fallback to commons-compress ZIP. But I don't know what this could be. Are you aware of missing features in the Java implementation that could be detected automatically?
Btw. they state by themselves:

> ZipArchiveInputStream is limited and may even return false contents in some cases, use ZipFile whenever possible. See [the ZIP documentation page](https://commons.apache.org/proper/commons-compress/zip.html#ZipArchiveInputStream_vs_ZipFile) for details. This limitation is a result of streaming data vs using random access and not a limitation of Compress' specific implementation.

https://commons.apache.org/proper/commons-compress/limitations.html#ZIP

Again, nothing against Apache Commons Compress as a whole. But as first option for reading a zip file I would see `java.util.zip.ZipFile`.

You may have seen the branch `open-file-performance`. I have to commit some changes in the content providers and then would merge that. If you have an idea how to ideally fallback to commons `ZipArchiveInputStream`, I am willing to incorporate it.

mehmet-karaman commented:
I've pulled your branch, let me check that. Maybe we can try to create a new facade to modify only the necessary places and keep the original implementation as much as possible.

mehmet-karaman commented:
@uvoigt
I’ve reviewed your implementation, and I believe it won’t be difficult to introduce a facade along with wrapper implementations for both ZIP implementations. This should help us minimize the required changes as much as possible.

I’m not entirely sure yet how this will impact performance, but I’ve already started working on the corresponding implementations. My starting point is your `open-file-performance` branch.

uvoigt commented:
I am still not convinced that the commons-compress zip implementation should be used to read a zip file.
You said that
> It wasn't possible to override some functionalities in the zip implementation of java. 

What exactly do you mean?
Actually, the commons-compress implementation is 100 times slower and it cannot detect the entry sizes within jars.

I am searching for an argument. To have a facade still not solves any issue.

mehmet-karaman commented:
For example, the method java.util.zip.ZipFile.Source.checkAndAddEntry(int, int) is private, and therefore its functionality cannot be overridden. If you still try to use it, it throws the exception "invalid CEN header (bad compression method: …)".

For supporting Zstd, using Apache Commons Compress was the only viable option.
My idea was to use Apache Commons Compress only for ZIP files that actually contain Zstd-compressed entries. For all other ZIP files, we would rely on the standard Java ZIP implementation (Facade should cover both impls). That should resolve all remaining issues — or am I missing something?

Regarding the other issue:

<img width="674" height="342" alt="image" src="https://github.com/user-attachments/assets/7fb04966-1a50-47f7-80b3-f5dff39870fd" />

I tested with a JAR file, and I can see the sizes correctly.
Do you have a specially prepared JAR file that behaves differently?
In any case, for JAR files we could even skip the Zstd check entirely and directly use the Java ZIP implementation.

uvoigt commented:
I looked up the `checkAndAddEntry` topic and it seems to be a JRE issue which has been very well analyzed here: https://connortumbleson.com/2023/07/31/invalid-cen-header/

Reading the comments, I suspect the method was added by a non-pro.

Using the system property `jdk.util.zip.disableZip64ExtraFieldValidation` might disable that feature. But the commenter of the above found that there is no effect in using `System.setProperty` because that property is read within a static initializer of `ZipFile`.

Therefore, I would suggest to simply try/catch that `java.util.zip.ZipException: Invalid CEN header (invalid zip64 extra data field size)` and fallback to commons-compress `ZipArchiveInputStream` in the catch clause.

I don't know when this occurs and which kind of zip archives produce that trouble. Maybe `detectStream` would be a good place.

uvoigt commented:
After experimenting with a prepared invalid file, I would say that the only task for the Commons-Compress Zip implementation is to write a Zip file.
When reading that invalid Zip file, both `java.util.zip.ZipFile` and `org.apache.commons.compress.archivers.zip.ZipFile` throw an Exception.
Both, `java.util.zip.ZipInputStream` and `org.apache.commons.compress.archivers.zip.ZipArchiveInputStream` do not throw.
If I really want to open and edit such a file, the fallback should go to the faster one which is `java.util.zip.ZipInputStream`.
If I want to modify it, it won't be saved the same way it was before. The invalid extra headers will be removed.

When saving, the old code opened a ZipFile once in `ZipRootNode`. That was for performance reasons but has been removed by this PR. It must be restored or else the editor will need ages to store a large file because a new File is opened for each `ZipNode`.
If `ZipFile` cannot be used here, all `ZipEntries` including their content must be read in advance using `ZipInputStream` because there is nothing like `ZipFile.getInputStream(zipEntry)` in that context.

mehmet-karaman commented:
@uvoigt I am currently preparing the zstd changes up on your latest master branch.. I've seen that minimal required is java 1.8, is there any problem to update this to java-17? 

Another question is the target platform configuration which depends on 2022-09, shouldn't we update this too a never version? The apache commons compress version 1.28.0 is the minimal version which contains the zstd implementation. So it would be good if we update to an orbit which contains at least that.

uvoigt commented:
@mehmet-karaman Please lets move our discussion to the PR containing the zstd changes.
